### ISSUE TITLE
Discuss: API change to display shows

### ISSUE THREAD

**Issue by labrys** [SickRage/sickrage-issues#96]
Currently, as per the API description the paused option for `/api/{key}/?cmd=shows` functions as follows:

`True to include paused shows, False otherwise`

I propose changing it to a tri-state value. Something like:
```
1 = Paused shows only
2 = Un-paused shows only
3, 0 or None = Both
```

**Comment by labrys** [comment]
Obviously, being a change to the API can have some serious implications, but I think this would add a functionality that currently is unavailable (to show only paused shows).   Related issue: #90

**Comment by phaseburn** [comment]
This would be a benefit for 3rd party tools, and would also restore the default behaviour of displaying both (which is what SickRage has classically done, even if it was not in line with the documentation) for all existing 3rd party tools, some which may not notice the API "fix" we made in #90 for quite some time (DroidBeard comes to mind on that front).

**Comment by miigotu** [comment]
Cant really change the accepted values for the existing api, as it breaks 3rd party tools which may or may not be maintained anymore.

We had planned on an api v2 a few months ago which should live along side of the original api, and had deferred it until we were more deperate for things to do =P

**Comment by labrys** [comment]
Kind of figured that... maybe we should start making a list of APIv2 changes for the future.

**Comment by Arcanemagus** [comment]
Sounds like a candidate for a milestone to track that :wink:.

**Comment by labrys** [comment]
Another candidate for a milestone: implement testing of all classes/modules/methods

**Comment by phaseburn** [comment]
@miigotu So I've been thinking a bit about this recently. We just *did* change the way this works, and, it did break a ton of 3rd party stuff. Even if it was a bug on our end, it was the "accepted" way the API worked for how long?

Look at the old behaviour (not the documentation). If you omitted a value entirely for "paused", it returned both paused and unpaused shows. If you sent a value of "paused=1", it would *only* display paused shows, so, nothing was *really* using it, despite the documentation saying it was necessary. Now, we've altered that behaviour so that paused=1 is *required* to show paused shows along with the unpaused shows. This broke every 3rd party app that listed shows, in that they weren't showing paused ones any more. If anybody was depending on it to return only paused shows, they're not getting what they expect any more, either. So, I'd suggest we redo the attribute so that omitting "paused" entirely defaults to a mixed list like we had before. Leave paused=1 the way it is, so that it, too, returns a list of paused and unpaused shows (as, that's what the documentation has always stated, but it never actually worked). Yes, anybody using the old behaviour (who is currently broken) will still be broken. But this will not break anything that has since updated to our "fix". It also restores the old behaviour and fixes the issues we've caused on projects that haven't updated (yet, or, otherwise). Then, lets add another two values for the "paused" attribute: perhaps not numeric, such as "only" to retrieve a listing of only paused shows, and, "none" to retrieve a list of only unpaused shows.

Basically, this change breaks nothing (that isn't already broken), *fixes most of the issues we've already caused*, and, gives the existing API a bit of functionality that the documentation has always said existed in the first place. Obviously, we'd have to update the doc to match the new values.

**Comment by fernandog** [comment]
@phaseburn NZB360, DroidBeard and ShowsRage are working with the actual code

What are the others third party apps ?

**Comment by phaseburn** [comment]
@fernandog Actually, DroidBeard and NZB360 do not work properly. ShowsRage does, and I submitted the PR myself that fixed it.

Open up DroidBeard and select a show. Pause it. Pause a few, if you want. Then, close DroidBeard. Maybe reboot your device if you want to. Anything that'll close DroidBeard, as in, you're coming back later to manage stuff. Open it back up. Now, lets say you want to unpause your show(s), but you'll find all the paused ones are not listed. Why? Because the default behavior in SickRage's API has changed to NOT including paused shows with out "paused=1" being sent as a parameter. And, those apps haven't been updated to include that parameter. Other SickBeard-compatible tools will have the same issue.

**Comment by fernandog** [comment]
I thought we were talking about **list shows** not the pause command
all shows are correctly listed those apps.
I can send you some screenshots.

check api builder explanation:
its all about include or not paused shows in the list

![image](https://cloud.githubusercontent.com/assets/2620870/11594197/bfad6062-9a8f-11e5-9ef7-4d0c19627d0b.png)

**Comment by phaseburn** [comment]
I understand what you are saying. I believe you are missing my point.

Prior to this being "fixed" by us, including "paused=true" would ONLY display paused shows. It never worked as documented. As such, nobody elected to use it. Then we "fixed" this, which changed the way the API has historically worked for a very long time. That caused everything to not display paused shows any more. We broke those tools with an API change. I'm suggesting we FIX that on our end, rather than requiring all these tools to update their API call to include "paused=true" on their ends.

**Comment by fernandog** [comment]
I understood. I remember SB api. But afaik SR changed that. 
Don't know when. 
maybe when @MGaetan89 did the api rewrite or before  that.

SB api:
paused (optional)
0 - show only non paused
1 - show only paused

**Comment by phaseburn** [comment]
SR changed it just last week. Maybe it was _supposed_ to be changed long before, but it was just last week that it went live. And because people have been using it the whole time as it always was, SR has just broken all of those tools. I'm merely suggesting we change it back to how it was, and unbreaking those tools which are either unlikely to update because that would make them incompatible with other forks of SickBeard, or, are no longer under active development.

**Comment by OmgImAlexis** [comment]
Just throwing me 2c in here but I honestly think if we need to break things to fix them then we might as well. There's no point in changing things back just so some 3rd party app that was using the broken code can work. If sickrage no longer works with those apps then either they need to update their apps or people will need to use ones that work with the updated API. Keep in mind large companies do this already with their APIs and yes developers and users get annoyed but eventually they update their apps or users move to apps that work. 

Anyways, back to sleep for me.

**Comment by Thraxis** [comment]
The API should work the way the API documentation says it should.  If it had not been working and developers using the API did not report it as broken, but instead used that broken API call as "Well it works this way" that doesn't make changing the documentation / API to reflect that broken implementation the right thing to do.

**Comment by labrys** [comment]
@OmgImAlexis generally when a developer makes a change (even just to fix something that was broken that people came to rely on working a certain way) they generally support both options for a period of time after deprecating a method.

@Thraxis if the api functions a certain way, and has for a long period of time, the documentation should change, not the code.

Having said all that, I think the api is in need of an overhaul (thus the APIv2 option).  I think the API development should start sooner rather than later.  It could be added to the sickrage/ folder while leaving the original SB code untouched.  Expose the new API at a new endpoint (such as /apiv2 or /api-beta or something)  so it can be tested as we flesh it out and so people who use it can begin to implement it, with the understanding that until its finalized it can (and likely will) change at any time.

**Comment by miigotu** [comment]
I think the original api should return to however it worked with the old (no longer maintained) 3rd party tools. If we were ok with breaking support for old tools there would be no reason for an apiv2 and we would just modify the api all willy nilly.

**Comment by labrys** [comment]
Closing this as it's already on the API v2 list in #99
